File: webadmin/security.py
"""文件访问路径安全校验（spec: web-admin-server）。

所有按相对路径读取文件的 API 必须经 resolve_safe_path：
规范化 + 白名单根目录约束，拒绝 `..` 穿越与符号链接逃逸。
"""

from pathlib import Path

from .settings import ALLOWED_FILE_ROOTS


class PathSecurityError(ValueError):
    """路径越界或非法"""


def resolve_safe_path(root_key: str, rel_path: str) -> Path:
    """把 (白名单根目录键, 相对路径) 解析为安全的绝对路径。

    - root_key 必须是 ALLOWED_FILE_ROOTS 之一（output/logs/flows/extracts）
    - rel_path 解析后必须仍位于对应根目录内（realpath 校验，符号链接也无法逃逸）
    """
    root = ALLOWED_FILE_ROOTS.get(root_key)
    if root is None:
        raise PathSecurityError(f"非法的根目录: {root_key}")

    rel = (rel_path or "").strip().lstrip("/")
    if "\x00" in rel:
        raise PathSecurityError("路径包含非法字符")

    candidate = (root / rel).resolve()
    root_resolved = root.resolve()
    if candidate != root_resolved and root_resolved not in candidate.parents:
        raise PathSecurityError(f"路径越界: {rel_path}")
    return candidate

File: apps/api/src/security.ts
import { existsSync, statSync } from "node:fs";
import path from "node:path";

import { badRequest } from "./errors.js";

export type AllowedRoot = "output" | "logs" | "flows" | "extracts" | "static";

export function resolveSafePath(
  roots: Record<AllowedRoot, string>,
  rootKey: AllowedRoot,
  relPath = ""
): string {
  const root = roots[rootKey];
  if (!root) {
    throw badRequest(`非法的根目录: ${rootKey}`);
  }
  let rel: string;
  try {
    rel = decodeURIComponent(relPath);
  } catch {
    rel = relPath;
  }
  rel = rel.trim().replace(/^[/\\]+/, "");
  if (rel.includes("\0")) {
    throw badRequest("路径包含非法字符");
  }
  const rootResolved = path.resolve(root);
  const candidate = path.resolve(rootResolved, rel);
  if (candidate !== rootResolved && !candidate.startsWith(`${rootResolved}${path.sep}`)) {
    throw badRequest(`路径越界: ${relPath}`);
  }
  return candidate;
}

export function isFile(filePath: string): boolean {
  return existsSync(filePath) && statSync(filePath).isFile();
}

export function isDirectory(filePath: string): boolean {
  return existsSync(filePath) && statSync(filePath).isDirectory();
}
